qa/L0_model_update/instance_update_test.py

```diff
@@ -0,0 +1,310 @@
+            self.__triton.infer("instance_init_del", self.__get_inputs((1, 4)))
+
+    def test_update_while_inferencing(self):
+        # Load model
+        update_instance_group("{\ncount: 1\nkind: KIND_CPU\n}")
+        self.__triton.load_model("instance_init_del")
+        self.assertEqual(get_initialize_count(), 1)
+        self.assertEqual(get_finalize_count(), 0)
+        self.__triton.infer("instance_init_del", self.__get_inputs((2,)))
+        # Add 1 instance while inferencing
+        set_infer_delay(10)
+        update_instance_group("{\ncount: 2\nkind: KIND_CPU\n}")
+        with concurrent.futures.ThreadPoolExecutor() as pool:
+            infer_start_time = time.time()
+            infer_thread = pool.submit(self.__triton.infer, "instance_init_del",
+                                       self.__get_inputs((1,)))
+            time.sleep(2)  # make sure inference has started
+            update_start_time = time.time()
+            update_thread = pool.submit(self.__triton.load_model,
+                                        "instance_init_del")
+            update_thread.result()
+            update_end_time = time.time()
+            infer_thread.result()
+            infer_end_time = time.time()
+        infer_time = infer_end_time - infer_start_time
+        update_time = update_end_time - update_start_time
+        # Adding a new instance does not depend on existing instances, so the
+        # ongoing inference should not block the update.
+        self.assertGreaterEqual(infer_time, 10.0, "Invalid infer time")
+        self.assertLess(update_time, 5.0, "Update blocked by infer")
+        self.assertEqual(get_initialize_count(), 2)
+        self.assertEqual(get_finalize_count(), 0)
+        self.__triton.infer("instance_init_del", self.__get_inputs((8,)))
+        # Unload model
+        self.__triton.unload_model("instance_init_del")
+        time.sleep(10)  # wait for unload to complete
+        self.assertEqual(get_initialize_count(), 2)
+        self.assertEqual(get_finalize_count(), 2)
+        with self.assertRaises(InferenceServerException):
+            self.__triton.infer("instance_init_del", self.__get_inputs((4,)))
+
+    def test_infer_while_updating(self):
+        # Load model
+        update_instance_group("{\ncount: 1\nkind: KIND_CPU\n}")
+        self.__triton.load_model("instance_init_del")
+        self.assertEqual(get_initialize_count(), 1)
+        self.assertEqual(get_finalize_count(), 0)
+        self.__triton.infer("instance_init_del", self.__get_inputs((2,)))
+        # Infer while adding 1 instance
+        set_initialize_delay(10)
+        update_instance_group("{\ncount: 2\nkind: KIND_CPU\n}")
+        with concurrent.futures.ThreadPoolExecutor() as pool:
+            update_start_time = time.time()
+            update_thread = pool.submit(self.__triton.load_model,
+                                        "instance_init_del")
+            time.sleep(2)  # make sure update has started
```

Comment:
Why? Didn't we want to test inference when update was in flight? I would assume a sequence of inference requests hitting the server when model instance update is in progress to be a better test.

Reply:
We test both: [test_update_while_inferencing](https://github.com/triton-inference-server/server/pull/5635/files#diff-40ca2af021f4288f3fc6ae6071a244bca6350a0e657ab2822b61650cac266931R230) and [test_infer_while_updating](https://github.com/triton-inference-server/server/pull/5635/files#diff-40ca2af021f4288f3fc6ae6071a244bca6350a0e657ab2822b61650cac266931R269)